numpy/testing/utils.py

```diff
@@ -1718,3 +1718,44 @@ def tempdir(*args, **kwargs):
     tmpdir = mkdtemp(*args, **kwargs)
     yield tmpdir
     shutil.rmtree(tmpdir)
+
+
+class catch_and_clear_warnings(warnings.catch_warnings):
+    """ ``catch_warnings`` context manager that resets warning registry
+
+    Warnings can be slippery, because, whenever a warning is triggered, Python
+    adds a ``__warningregistry__`` member to the *calling* module.  This makes
+    it impossible to retrigger the warning in this module, whatever you put in
+    the warnings filters.  This context manager accepts a sequence of `modules`
+    as a keyword argument to its constructor and:
+
+    * stores and removes any ``__warningregistry__`` entries in given `modules`
+      on entry;
+    * resets ``__warningregistry__`` to its previous state on exit.
+
+    This makes it possible to trigger any warning afresh inside the context
+    manager without disturbing the state of warnings outside.
```

Comment:
Need to document the arguments for the constructor here. An example showing how to get the module name would be good.